### PR TITLE
8334010: VM crashes with ObjectAlignmentInBytes > GCCardSizeInBytes

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  please review this change that makes VM initialization reject invalid combinations of `ObjectAlignmentInBytes` and `GCCardSizeInBytes`.

ObjectAlignmentInBytes larger than GCCardSizeInBytes causes issues with setting correct BOT offsets.

Testing: new test case, gha

Hth,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334010](https://bugs.openjdk.org/browse/JDK-8334010): VM crashes with ObjectAlignmentInBytes &gt; GCCardSizeInBytes (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21515/head:pull/21515` \
`$ git checkout pull/21515`

Update a local copy of the PR: \
`$ git checkout pull/21515` \
`$ git pull https://git.openjdk.org/jdk.git pull/21515/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21515`

View PR using the GUI difftool: \
`$ git pr show -t 21515`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21515.diff">https://git.openjdk.org/jdk/pull/21515.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21515#issuecomment-2413199511)

### CODE REVIEW

File: test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        "-Xmx32m",
+        "-Xms32m",
+        "-version");
+
+    System.out.println("Output:\n" + output.getOutput());
+
+    if (shouldSucceed) {
+      output.shouldHaveExitValue(0);
+    } else {
+      output.shouldContain("Error occurred during initialization of VM");
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java:51]: Check for `Invalid combination of GCCardSizeInBytes and ObjectAlignmentInBytes` message explicitly?

File: test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class TestObjectAlignmentCardSize {
+  private static void runTest(int objectAlignment, int cardSize, boolean shouldSucceed) throws Exception {
+    OutputAnalyzer output = ProcessTools.executeTestJava(
+        "-XX:ObjectAlignmentInBytes=" + objectAlignment,
+        "-XX:GCCardSizeInBytes=" + cardSize,
+        "-Xmx32m",
+        "-Xms32m",
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java:43]: I wonder if we are playing close to heap limits here, especially with large `ObjectAlignmentInBytes`. Do you have a feeling how much of the heap is used with `ObjectAlignmentInBytes=512`? Should this be `128m` instead, maybe?

**Review by shipilev — COMMENTED** [review]
This looks fine, I only have comments about the test.

File: test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java (hunk 2 above)
Comment (tschatzl) [reply test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java:43]: The gc+heap+exit memory usage shows <5m memory usage for all GCs (G1, Shenandoah, Serial, Parallel). We could go even lower to 16m easily.

Comment (shipilev) [reply test/hotspot/jtreg/gc/TestObjectAlignmentCardSize.java:43]: Cool, nevermind then.

**Review by shipilev — APPROVED** [review]

**Review by walulyai — APPROVED** [review]
LGTM!

**Comment by tschatzl** [comment]
Thanks @shipilev @walulyai for your review
/integrate

**Pull request closed without merging**